lib/using_restricted_array.rb

```diff
@@ -6,56 +6,150 @@
 end
 
 # Prints each integer values in the array
-# Time complexity: ?
-# Space complexity: ?
+# Time complexity: O(n)
+# Space complexity: O(1)
 def print_array(array)
-  raise NotImplementedError
+  printed_elements = ""
+  n = 0
+
+  until array[n] == nil
+    printed_elements += "#{array[n]} "
+    n += 1
+  end
+  #raise NotImplementedError
+  return printed_elements
 end
 
 # For an unsorted array, searches for 'value_to_find'.
 # Returns true if found, false otherwise.
-# Time complexity: ?
-# Space complexity: ?
+# Time complexity:  O(n)
+# Space complexity: O(1)
 def search(array, length, value_to_find)
-  raise NotImplementedError
+  n = 0
+  value_search = false
+
+  until n == length
+    if array[n] == value_to_find
+      value_search = true
+    end
+    n += 1
+  end
+
+  return value_search
+
+  #raise NotImplementedError
 end
 
 # Finds and returns the largest integer value the array
 # Assumes that the array is not sorted.
-# Time complexity: ?
-# Space complexity: ?
+# Time complexity: O(n)
+# Space complexity: O(1)
 def find_largest(array, length)
-  raise NotImplementedError
+  largest_value = 0
+  n = 0
+
+  until n == length
+    if array[n] == nil
+      return false
+    elsif array[n] > largest_value
+      largest_value = array[n]
+    end
+    n += 1
+  end
+
+  return largest_value
+  #raise NotImplementedError
 end
 
 # Finds and returns the smallest integer value in the array
 # Assumes that the array is not sorted.
-# Time complexity: ?
-# Space complexity: ?
+# Time complexity: O(n)
+# Space complexity: O(1)
 def find_smallest(array, length)
-  raise NotImplementedError
+  smallest_value = 221
+  n = 0
+
+  until n == length
+    if array[n] == nil
+      return false
+    elsif array[n] < smallest_value
+      smallest_value = array[n]
+    end
+
+    n+=1
+
+  end
+
+  return smallest_value
+
+  #raise NotImplementedError
 end
 
 # Reverses the values in the integer array in place
-# Time complexity: ?
-# Space complexity: ?
+# Time complexity: O(n)
+# Space complexity: O(1)
 def reverse(array, length)
-  raise NotImplementedError
+  i = 0
+  j = (length - 1)
+
+  if length <= 1
+    return array
+  end
+
+  while i < j
+    temp = array[i]
+    array[i] = array[j]
+    array[j] = temp
+
+    i += 1
+    j -= 1
+  end
+
+  return array
+  # raise NotImplementedError
 end
 
 # For an array sorted in ascending order, searches for 'value_to_find'.
 # Returns true if found, false otherwise.
-# Time complexity: ?
-# Space complexity: ?
+# Time complexity: O(log(n))
+# Space complexity: O(n)
 def binary_search(array, length, value_to_find)
-  raise NotImplementedError
+  low = 0
+  high = length - 1
+
+  if length == 0
+    return false
+  end
+
+  while low <= high
+    mid = (low + high) / 2
+    if array[mid] == value_to_find
+      return true
+    elsif array[mid] > value_to_find
+      high = mid - 1
+    else
+      low = mid + 1
+    end
+  end
+
+  if array[mid] != value_to_find
+   return false
+  end
```

Comment:
Just a little boolean zen

```suggestion
  return array[mid] == value_to_find
```